### PR TITLE
Router fix

### PR DESCRIPTION
**Pull request by BondPV**
fix router;
add search field

### CODE REVIEW

**Review by Alesia-V175 — APPROVED** [review]

File: src/index.html [hunk 1]
```diff
@@ -38,27 +38,25 @@
...
-              <a href="#" class="nav-menu__item-link">Tablets</a>
+              <a href="#?category=Tablets&brand=&price=&stock=&search=&sort=&view=" class="nav-menu__item-link">Tablets</a>
             </li>
             <li class="nav-menu__item">
-              <a href="#" class="nav-menu__item-link">Game Consoles</a>
+              <a href="#?category=Consoles&brand=&price=&stock=&search=&sort=&view=" class="nav-menu__item-link">Game Consoles</a>
             </li>
             <li class="nav-menu__item">
-              <a href="#" class="nav-menu__item-link">TV</a>
+              <a href="#?category=TV&brand=&price=&stock=&search=&sort=&view=" class="nav-menu__item-link">TV</a>
```
Comment (Ksarelto) [inline src/index.html]: u copypast almost the same stirng in several places, u can make some method or addition ulit function that will create to u this string

File: src/helpers/router/UrlHash.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      return;
+    } else {
+      this.setUrlHash(this.hashData);
+    }
+  }
+
+  static clearHash() {
+    window.location.hash = '#main';
+    this.hashData = {
+      filtersValue: { category: [], brand: [] },
```
Comment (Ksarelto) [inline src/helpers/router/UrlHash.ts]: object values usually write from new line, for better readability
```
filtersValue: {
  category: [],
  brand: []
}
```

File: src/helpers/router/UrlHash.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    const currentHash = decodeURI(window.location.hash).slice(2);
+    this.getHashData(currentHash);
+    return this.hashData[param];
+  }
+
+  public static setUrlHashParam(param: FiltersName.Sort | FiltersName.Search | FiltersName.View, value: string): void {
+    this.hashData[param] = value;
+
+    if (param === FiltersName.Sort && value === '') {
+      return;
```
Comment (Ksarelto) [inline src/helpers/router/UrlHash.ts]: function return udefined by default, so no need to this by yourself, better create condition that will call necessary logic, and return nothig
```
setUrlHashParam() {
  this.hashData[param] = value;
  
  if (!value === '') {
   this.setUrlHash(this.hashData)
  }
}
```

File: src/helpers/router/UrlHash.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    for (let i = 0; i < urlHashArray.length; i++) {
+      const param: string[] = urlHashArray[i].split('=');
+
+      if ((param[0] === FiltersName.Category || param[0] === FiltersName.Brand) && param[1].length > 0) {
+        hashData.filtersValue[param[0]] = param[1].split('+');
+        hashData.filtersValue = hashData.filtersValue;
+      }
+
+      if ((param[0] === FiltersName.Price || param[0] === FiltersName.Stock) && param[1].length > 0) {
```
Comment (Ksarelto) [inline src/helpers/router/UrlHash.ts]: if many conditions, better to write them in multiline
```
if (
   (param[0] === FiltersName.Price || param[0] === FiltersName.Stock) &&
   param[1].length > 0
)

File: src/helpers/router/UrlHash.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    window.location.hash = `?${category}&${brand}&${price}&${stock}&${search}&${sort}&${view}`;
+  }
+
+  static getHashData(hash: string): HashDataType {
+    const hashData = this.hashData;
+    const urlHashArray: string[] = hash.slice(1).split('&');
+
+    for (let i = 0; i < urlHashArray.length; i++) {
+      const param: string[] = urlHashArray[i].split('=');
```
Comment (Ksarelto) [inline src/helpers/router/UrlHash.ts]: destruct param and use its values
```
const [firstParam, secondParam] = urlHashArray[i].split('=');
```

File: src/helpers/router/Router.ts [hunk 6]
```diff
@@ -28,20 +30,17 @@ class Router {
   private setPage(hash: string) {
     if (window.location.hash.length === 0) {
       window.location.hash = `#${Pages.Main}`;
-    }
-
-    if (hash.includes(Pages.ProductDetails)) {
+    } else if (hash === Pages.Main) {
```
Comment (Ksarelto) [inline src/helpers/router/Router.ts:33]: can u use here 'switch case'?

File: src/helpers/router/Router.ts [hunk 7]
```diff
@@ -28,20 +30,17 @@ class Router {
...
-
-    switch (hash) {
-      case Pages.Main:
-        this.currentPage = new MainPage(Container);
-        break;
-
-      case Pages.Cart:
-        this.currentPage = new CartPage(Container);
-        break;
+    } else if (hash.includes('?')) {
```
Comment (Ksarelto) [inline src/helpers/router/Router.ts]: for '+', '&', '?' create an enum SpecialSymbols, and use it insuch cases

File: src/components/main/grid/Grid.ts [hunk 8]
```diff
@@ -45,8 +47,22 @@ class Grid {
     viewButtonList.classList.add(ClassListName.catalogView, ClassListName.catalogViewList);
     catalogHeader.append(viewButtonList);
 
-    viewButtonList.addEventListener('click', () => this.changeGridList());
-    viewButtonTable.addEventListener('click', () => this.changeGridTable());
+    const currentGridValue = UrlHash.getUrlHashParam(FiltersName.View);
+
+    if (currentGridValue === 'list') {
```
Comment (Ksarelto) [inline src/components/main/grid/Grid.ts]: this string is used in several places, create a constant for it

File: src/components/main/sortBar/filters/Filters.ts [hunk 9]
```diff
@@ -104,22 +105,25 @@ class Filters {
...
     }
-
     this.filterProducts();
     this.appendTotalFoundQuantity();
   }
 
   //* render Catalog
   public filterProducts(): void {
-    const savedFilters: FiltersDataType = JSON.parse(localStorage.getItem('filters') as string);
+    const savedFilters: FiltersDataType = UrlHash.hashData;
```
Comment (Ksarelto) [inline src/components/main/sortBar/filters/Filters.ts]: can u just destruct UrlHashData?
```
const { filtersValues, filtersRange, search ] = UrlHash.hashData
```

File: src/components/main/sortBar/filters/Filters.ts [hunk 10]
```diff
@@ -224,10 +228,10 @@ class Filters {
     return items.filter((item) => item[filterName] >= filterRange[0] && item[filterName] <= filterRange[1]);
   }
 
-  // check for activity on saved from Local Storage
+  // check for activity on saved
   checkActiveElement(elementValue: string, element: HTMLInputElement): void {
```
Comment (Ksarelto) [inline src/components/main/sortBar/filters/Filters.ts]: method modifier

File: src/components/main/sortBar/filters/Filters.ts [hunk 11]
```diff
@@ -261,31 +265,58 @@ class Filters {
...
+      return (
+        titleDatail.includes(currentInput.toLowerCase()) ||
+        category.includes(currentInput.toLowerCase()) ||
+        stock.includes(currentInput) ||
+        price.includes(currentInput) ||
+        description.includes(currentInput.toLowerCase())
+      );
+    });
+
+    return (currentInput = '') ? filteredProducts : newProducts;
```
Comment (Ksarelto) [inline src/components/main/sortBar/filters/Filters.ts]: better to check for empty input in the start of the method
```
if (currentInput === '') {
  return filteredProducts
}

return filteredProducts.filter(item)
```

File: src/helpers/router/UrlHash.ts (hunk 2 above)
Comment (BondPV) [reply src/helpers/router/UrlHash.ts]: fixed

File: src/helpers/router/UrlHash.ts (hunk 3 above)
Comment (BondPV) [reply src/helpers/router/UrlHash.ts]: fixed

File: src/index.html (hunk 1 above)
Comment (BondPV) [reply src/index.html]: fixed

File: src/helpers/router/UrlHash.ts (hunk 4 above)
Comment (BondPV) [reply src/helpers/router/UrlHash.ts]: Error when wrapping a line
Delete `⏎······`eslint [prettier/prettier](https://github.com/prettier/eslint-plugin-prettier#options).

TODO. Need to figure out how to set up prettier

File: src/helpers/router/UrlHash.ts (hunk 5 above)
Comment (BondPV) [reply src/helpers/router/UrlHash.ts]: fixed

File: src/helpers/router/Router.ts (hunk 7 above)
Comment (BondPV) [reply src/helpers/router/Router.ts]: added enum Symbol

File: src/components/main/grid/Grid.ts (hunk 8 above)
Comment (BondPV) [reply src/components/main/grid/Grid.ts]: added constant

File: src/components/main/sortBar/filters/Filters.ts (hunk 9 above)
Comment (BondPV) [reply src/components/main/sortBar/filters/Filters.ts]: fixed

File: src/components/main/sortBar/filters/Filters.ts (hunk 11 above)
Comment (BondPV) [reply src/components/main/sortBar/filters/Filters.ts]: fixed

File: src/components/main/sortBar/filters/Filters.ts (hunk 10 above)
Comment (BondPV) [reply src/components/main/sortBar/filters/Filters.ts]: added metod modifier

**Pull request merged**